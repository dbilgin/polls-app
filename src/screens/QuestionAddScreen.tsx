import React, {useState} from 'react';
import {
  StyleSheet,
  View,
  Alert,
  TextInput,
  ScrollView,
  TouchableOpacity,
  Text,
} from 'react-native';
import {useDispatch} from 'react-redux';
import {RootStackParamList} from '../../App';
import {StackNavigationProp} from '@react-navigation/stack';
import {addQuestion} from '../../PollsService';
import {debounce} from 'lodash';
import {FAB} from 'react-native-paper';
import {fetchQuestionSuccess} from '../store/actions';
import {Question} from '../../Models/Question';

type QuestionAddScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'QuestionAddScreen'
>;

type Props = {
  navigation: QuestionAddScreenNavigationProp;
};

const QuestionAddScreen: React.FC<Props> = ({navigation}) => {
  const [title, setTitle] = useState('');
  const [choices, setChoices] = useState<string[]>([]);
  const dispatch = useDispatch();

  const cleanChoices = () => {
    return choices.filter((choice) => !!choice);
  };

  const add = async () => {
    const cleanedChoices = cleanChoices();

    if (
      !!title &&
      !!cleanedChoices &&
      !!cleanedChoices.length &&
      cleanedChoices.length >= 2
    ) {
      try {
        const addResult = await addQuestion(title, cleanedChoices);
        if (addResult.status === 201) {
          let newQuestion: Question = addResult.data;
          newQuestion.id = Number(newQuestion.url.split('/').pop());
          dispatch(fetchQuestionSuccess(newQuestion, newQuestion.id));
          navigation.goBack();
        } else {
          Alert.alert('An error has occurred.');
        }
      } catch {
        Alert.alert('An error has occurred.');
      }
    } else {
      Alert.alert('Enter all the fields.');
    }
  };

  const debounceSetTitle = debounce((text: string) => {
    setTitle(text);
  }, 500);

  const computeChoice = (choice: string, index: number) => {
    const newChoices = [...choices];
    newChoices[index] = choice;
    setChoices(newChoices);
  };

  const addChoice = () => {
    const newChoices = [...choices];
    newChoices.push('');
    setChoices(newChoices);
  };

  return (
    <View style={styles.container}>
      <ScrollView>
        <View style={{padding: 24}}>
          <TextInput
            style={styles.input}
            onChangeText={(text) => debounceSetTitle(text)}
          />
          <Text style={styles.titleText}>Choices</Text>
          {choices.map((__: string, index: number) => (
            <TextInput
              key={index.toString()}
              style={styles.input}
              onChangeText={(text) => computeChoice(text, index)}
            />
          ))}
          <TouchableOpacity style={{marginTop: 18}} onPress={addChoice}>
            <View style={styles.addChoiceButton}>
              <Text style={styles.buttonText}>Add New Choice</Text>
            </View>
          </TouchableOpacity>
        </View>
      </ScrollView>

      <FAB style={styles.fab} icon="content-save" onPress={() => add()} />
    </View>
  );
};

export default QuestionAddScreen;

const styles = StyleSheet.create({
  fab: {
    position: 'absolute',
    margin: 30,
    right: 0,
    bottom: 0,
    backgroundColor: '#147efb',
  },
  addChoiceButton: {
    height: 50,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#147efb',
  },
  buttonText: {
    fontSize: 16,
    color: 'white',
  },
  titleText: {
    fontSize: 24,
    color: '#505b7d',
    marginTop: 18,
  },
  input: {
    marginTop: 8,
    height: 40,
    borderColor: 'gray',
    borderWidth: 1,
    borderRadius: 16,
    paddingLeft: 12,
    paddingRight: 12,
  },
  choiceText: {
    fontSize: 18,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: 'white',
  },
  bottomText: {
    fontSize: 24,
    color: 'black',
    marginTop: 24,
  },
  dateText: {
    fontSize: 16,
    color: '#505b7d',
    marginTop: 12,
  },
  subtitle: {
    fontSize: 12,
    color: '#505b7d',
    marginTop: 12,
  },
  title: {
    fontSize: 24,
    color: '#505b7d',
  },

  cardContainer: {
    margin: 18,
    height: 100,
    borderWidth: 0,
    borderRadius: 18,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,

    elevation: 8,
  },
});
