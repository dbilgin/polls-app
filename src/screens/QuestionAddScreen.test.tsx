import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import QuestionAddScreen from './QuestionAddScreen';
import {TouchableOpacity, TextInput} from 'react-native';

Enzyme.configure({adapter: new Adapter()});

const createTestProps = () => ({
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
  },
});

jest.mock('../PollsService', () => ({
  addQuestion: jest.fn(() =>
    Promise.resolve({status: 201, data: {url: 'url/1'}}),
  ),
}));

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));

describe('QuestionAddScreen', () => {
  describe('Rendering', () => {
    let props: any;
    beforeEach(() => {
      props = createTestProps();
    });

    it('should match to snapshot', () => {
      const component = shallow(<QuestionAddScreen {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});

describe('Interaction', () => {
  let props: any;
  beforeEach(() => {
    props = createTestProps();
  });

  it('should call add new choice field', () => {
    const component = shallow(<QuestionAddScreen {...props} />);
    expect(component.find('TextInput').length).toBe(1);
    // @ts-ignore
    component.find(TouchableOpacity).first().props().onPress();
    expect(component.find('TextInput').length).toBe(2);
  });

  it('should not save without filling in', () => {
    const component = shallow(<QuestionAddScreen {...props} />);

    // @ts-ignore
    component.find(TouchableOpacity).at(1).props().onPress();
    expect(props.navigation.goBack).not.toHaveBeenCalled();
  });

  it('adding process should work', async () => {
    const component = shallow(<QuestionAddScreen {...props} />);

    // @ts-ignore
    await component.find(TouchableOpacity).first().props().onPress();
    // @ts-ignore
    await component.find(TouchableOpacity).first().props().onPress();

    component.find(TextInput).at(0).simulate('changeText', 'Test Test');
    component.find(TextInput).at(1).simulate('changeText', 'Test Test');
    component.find(TextInput).at(2).simulate('changeText', 'Test Test');

    expect(component.find(TouchableOpacity).length).toBe(2);
    expect(component.find(TextInput).length).toBeGreaterThan(2);
  });
});
