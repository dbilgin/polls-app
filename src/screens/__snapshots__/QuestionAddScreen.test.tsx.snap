// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionAddScreen Rendering should match to snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ScrollViewMock>
    <View
      style={
        Object {
          "padding": 24,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "gray",
            "borderRadius": 16,
            "borderWidth": 1,
            "height": 40,
            "marginTop": 8,
            "paddingLeft": 12,
            "paddingRight": 12,
          }
        }
        underlineColorAndroid="transparent"
      />
      <Text
        style={
          Object {
            "color": "#505b7d",
            "fontSize": 24,
            "marginTop": 18,
          }
        }
      >
        Choices
      </Text>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "marginTop": 18,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#147efb",
              "height": 50,
              "justifyContent": "center",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 16,
              }
            }
          >
            Add New Choice
          </Text>
        </View>
      </TouchableOpacity>
    </View>
  </ScrollViewMock>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "#147efb",
        "borderRadius": 50,
        "borderWidth": 0,
        "bottom": 0,
        "margin": 30,
        "padding": 15,
        "position": "absolute",
        "right": 0,
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/save.png",
        }
      }
      style={
        Object {
          "height": 40,
          "tintColor": "white",
          "width": 40,
        }
      }
    />
  </TouchableOpacity>
</View>
`;
